extra_data/components.py

```diff
@@ -926,6 +738,311 @@ def zip_trains_pulses(trains, pulses):
     return res
 
 
+
+class MultimodKeyData:
+    def __init__(self, det: MultimodDetectorBase, key):
+        self.det = det
+        self.key = key
+        self.modno_to_keydata = {
+            m: det.data[s, key] for (m, s) in det.modno_to_source.items()
+        }
+
+    @property
+    def train_ids(self):
+        return self.det.train_ids
+
+    def train_id_coordinates(self):
+        return self.det.train_ids
+
+    @property
+    def modules(self):
+        return sorted(self.modno_to_keydata)
+
+    @property
+    def _eg_keydata(self):
+        return self.modno_to_keydata[min(self.modno_to_keydata)]
+
+    @property
+    def ndim(self):
+        return self._eg_keydata.ndim + 1
+
+    @property
+    def shape(self):
+        return ((len(self.modno_to_keydata), len(self.train_id_coordinates()))
+                + self._eg_keydata.entry_shape)
+
+    @property
+    def dimensions(self):
+        return ['module', 'trainId'] + ['dim_%d' % i for i in range(self.ndim - 2)]
+
+    @property
+    def dtype(self):
+        return self._eg_keydata.dtype
+
+    def _with_selected_det(self, det_selected):
+        # Overridden for XtdfImageMultimodKeyData to preserve pulse selection
+        return MultimodKeyData(det_selected, self.key)
+
+    def select_trains(self, trains):
+        return self._with_selected_det(self.det.select_trains(trains))
+
+    def split_trains(self, parts=None, trains_per_part=None, frames_per_part=None):
+        for det_split in self.det.split_trains(parts, trains_per_part, frames_per_part):
+            yield self._with_selected_det(det_split)
+
+    def ndarray(self, *, fill_value=None, out=None, roi=(), astype=None, module_gaps=False):
```

Comment:
I remember I originally also wanted to add an `out=` parameter to `components` in https://github.com/European-XFEL/EXtra-data/pull/307, but exposing the expected shape turned out not as trivial.

Looking at this now, it seems fairly straightforward and should only require `module_gaps` and `roi`. This could it significantly easier to pre-allocate arrays.

EDIT: It turns out `XtdfImageMultimodKeyData` actually has that method `_shape()` albeit private.

Reply:
They have `.shape` as a property, which is equivalent to `module_gaps=False` and no ROI. This matches KeyData objects, and various array-like things. `._shape()` is a private method just so I didn't have to come up with a good name for it - but if we agree on one, I'm happy to make it public.

Reply:
Oh, it's _only_ the naming bikeshed.

Maybe using the word buffer would make the intent clear?
`.buffer_shape(roi=(), module_gaps=False)`?
`.shape_for_buffer(...)?`

Reply:
I've called this `buffer_shape` at your suggestion - I don't have a strong feeling about the name.